substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/MemoryUtil.java

```diff
@@ -0,0 +1,211 @@
+                         */
+                        "-XX:MaxRAMPercentage=" + determineReasonableMaxRAMPercentage(),
+                        /*
+                         * Optimize for throughput by increasing the goal of the total time for
+                         * garbage collection from 1% to 5% (N=19). This also reduces peak RSS.
+                         */
+                        "-XX:GCTimeRatio=19", // 1/(1+N) time for GC
+                        /*
+                         * Let builder exit on first OutOfMemoryError to provide for shorter
+                         * feedback loops.
+                         */
+                        "-XX:+ExitOnOutOfMemoryError");
+    }
+
+    /**
+     * Returns a percentage (0.0-100.0) to be used as a value for the -XX:MaxRAMPercentage flag of
+     * the builder process. Prefer available memory over total memory to reduce memory pressure on
+     * the host machine.
+     */
+    private static double determineReasonableMaxRAMPercentage() {
+        var osBean = (com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
+        long totalMemorySize = osBean.getTotalMemorySize();
+        long reasonableMaxMemorySize = -1;
+        reasonableMaxMemorySize = switch (OS.getCurrent()) {
+            case LINUX -> getAvailableMemorySizeLinux();
+            case DARWIN -> getAvailableMemorySizeDarwin();
+            case WINDOWS -> getAvailableMemorySizeWindows();
+        };
+        if (reasonableMaxMemorySize < 0 || reasonableMaxMemorySize > totalMemorySize) {
+            return FALLBACK_MAX_RAM_PERCENTAGE;
+        }
+        if (reasonableMaxMemorySize < MIN_HEAP_BYTES) {
+            throw new NativeImageError(
+                            "There is not enough memory available on the system (got %sMiB, need at least %sMiB). Consider freeing up memory if builds are slow, for example, by closing applications that you do not need."
+                                            .formatted(reasonableMaxMemorySize / MiB_TO_BYTES, MIN_HEAP_BYTES / MiB_TO_BYTES),
+                            null, ExitStatus.OUT_OF_MEMORY.getValue());
+        }
+        reasonableMaxMemorySize = Math.min(reasonableMaxMemorySize, MAX_HEAP_BYTES);
+        return (double) reasonableMaxMemorySize / totalMemorySize * 100;
+    }
+
+    /**
+     * Returns the total amount of available memory in bytes on Linux based on
+     * <code>/proc/meminfo</code>, otherwise <code>-1</code>.
+     *
+     * @see <a href=
+     *      "https://github.com/torvalds/linux/blob/865fdb08197e657c59e74a35fa32362b12397f58/mm/page_alloc.c#L5137">page_alloc.c#L5137</a>
+     */
+    private static long getAvailableMemorySizeLinux() {
+        try {
+            String memAvailableLine = Files.readAllLines(Paths.get("/proc/meminfo")).stream().filter(l -> l.startsWith("MemAvailable")).findFirst().orElse("");
+            Matcher m = Pattern.compile("^MemAvailable:\\s+(\\d+) kB").matcher(memAvailableLine);
```

Comment:
Note that this metric is not container aware. I.e. on large systems, say with 1TB of memory, and a container slice of say 50GB you'd still get > 32 Gib since `50/0.8 > 32`. Using the OSBean could solve this problem.

Reply:
Good point and yes, in that case `reasonableMaxMemorySize > totalMemorySize` is true, so this would return the 80% fallback which actually matches the current behavior (80% of total memory). OSBean takes cgroups into account but it unfortunately only provides access to free, not available memory. Do you have any other suggestions how we could handle such cases?

Reply:
Following up on this in https://github.com/oracle/graal/pull/6581.